rest/data/generate.go

```diff
@@ -3,81 +3,45 @@ package data
 import (
 	"context"
 	"encoding/json"
-	"net/http"
-	"time"
 
-	"github.com/evergreen-ci/evergreen/model"
-	"github.com/evergreen-ci/evergreen/util"
-	"github.com/evergreen-ci/evergreen/validator"
-	"github.com/evergreen-ci/gimlet"
-	"github.com/mongodb/grip"
+	"github.com/evergreen-ci/evergreen/units"
+	"github.com/mongodb/amboy"
 	"github.com/pkg/errors"
-	mgo "gopkg.in/mgo.v2"
 )
 
 type GenerateConnector struct{}
 
 // GenerateTasks parses JSON files for `generate.tasks` and creates the new builds and tasks.
-func (gc *GenerateConnector) GenerateTasks(ctx context.Context, taskID string, jsonBytes []json.RawMessage) error {
-	projects, err := ParseProjects(jsonBytes)
-	if err != nil {
-		return gimlet.ErrorResponse{
-			StatusCode: http.StatusBadRequest,
-			Message:    errors.Wrap(err, "error parsing JSON from `generate.tasks`").Error(),
-		}
-	}
-	g := model.MergeGeneratedProjects(projects)
-	g.TaskID = taskID
-
-	_, err = util.Retry(
-		ctx,
-		func() (bool, error) {
-			p, v, t, pm, prevConfig, err := g.NewVersion() // nolint
-			if err != nil {
-				return false, err
-			}
-			if t.GeneratedTasks {
-				return false, nil // already generated tasks, noop
-			}
-			if err = validator.CheckProjectConfigurationIsValid(p); err != nil {
-				return false, err
-			}
-			err = g.Save(p, v, t, pm, prevConfig)
-			if err != nil && errors.Cause(err) == mgo.ErrNotFound {
-				return true, gimlet.ErrorResponse{
-					StatusCode: http.StatusInternalServerError,
-					Message:    errors.Wrap(err, "error updating config in `generate.tasks`").Error(),
-				}
-			}
-			if err = t.MarkGeneratedTasks(); err != nil {
-				return true, gimlet.ErrorResponse{
-					StatusCode: http.StatusInternalServerError,
-					Message:    errors.Wrapf(err, "problem marking task '%s' as having generated tasks", t.Id).Error(),
-				}
-			}
-			return false, nil
-		}, 10, 100*time.Millisecond)
-	return err
+func (gc *GenerateConnector) GenerateTasks(ctx context.Context, taskID string, jsonBytes []json.RawMessage, q amboy.Queue) error {
+	return q.Put(units.NewGenerateTaskJob(taskID, jsonBytes))
```

Comment:
Should use a concatenation of ID and type in case this queue is used for other things in the future

Reply:
`NewGenerateTaskJob` creates a job with type name.